### PR TITLE
App start: Add spans to first transaction

### PR DESCRIPTION
**Pull request by buenaflor**
## :scroll: Description

Adds app start spans to the first transaction

Added spans:
- App launch to plugin registration -> measures app launch until the plugin is attached to the flutter engine
- Main isolate setup -> measures time it takes from plugin registration until the earliest point dart is executed (as early as the flutter sdk is initialized)
- First frame render -> measures time between main isolate dart code executing until the first frame

## :bulb: Motivation and Context

mobile starfish

In combination with sentry navigator observer you'll be able to see app start traces in the app start module

## :green_heart: How did you test it?
Unit Tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: flutter/lib/src/sentry_flutter.dart [hunk 1]
```diff
@@ -43,6 +48,9 @@ mixin SentryFlutter {
   }) async {
     final flutterOptions = SentryFlutterOptions();
 
+    // ignore: invalid_use_of_internal_member
+    mainIsolateStartTime = flutterOptions.clock();
```
Comment (stefanosiano) [inline flutter/lib/src/sentry_flutter.dart]: why are you setting it here?
This changes its value, and it shouldn't, as far as i understand
This is why the value in the `AppStartInfo` is different from the one retrieved in `NativeAppStartEventProcessor`

File: flutter/lib/src/event_processor/native_app_start_event_processor.dart [hunk 2]
```diff
@@ -25,6 +33,82 @@ class NativeAppStartEventProcessor implements EventProcessor {
...
+        startTimestamp: appStartInfo.pluginRegistration,
+        endTimestamp: appStartInfo.mainIsolateStart);
+
+    final firstFrameRenderSpan = await _createAndFinishSpan(
+        tracer: transaction,
+        operation: appStartInfo.appStartTypeOperation,
+        description: appStartInfo.firstFrameRenderDescription,
+        parentSpanId: appStartSpan.context.spanId,
+        traceId: transactionTraceId,
+        startTimestamp: SentryFlutter.mainIsolateStartTime,
```
Comment (stefanosiano) [inline flutter/lib/src/event_processor/native_app_start_event_processor.dart]:
```suggestion
        startTimestamp: appStartInfo.mainIsolateStart,
```

File: flutter/lib/src/event_processor/native_app_start_event_processor.dart [hunk 3]
```diff
@@ -25,6 +33,82 @@ class NativeAppStartEventProcessor implements EventProcessor {
...
+      firstFrameRenderSpan
+    ]);
+  }
+
+  Future<SentrySpan> _createAndFinishSpan({
+    required SentryTracer tracer,
+    required String operation,
+    required String description,
+    required SpanId? parentSpanId,
+    required SentryId? traceId,
```
Comment (stefanosiano) [inline flutter/lib/src/event_processor/native_app_start_event_processor.dart]:
```suggestion
    required SpanId parentSpanId,
    required SentryId traceId,
```

File: flutter/lib/src/sentry_flutter.dart (hunk 1 above)
Comment (buenaflor) [reply flutter/lib/src/sentry_flutter.dart]: I think we can remove it.

`static DateTime mainIsolateStartTime = getUtcDateTime();` should be set as soon as SentryFlutter class is loaded, I guess it should be good enough to not have to use `flutterOptions.clock();` within init,

just tested it and there is around a 2-4ms difference

Comment (stefanosiano) [reply flutter/lib/src/sentry_flutter.dart]: the main problem is if `SentryFlutter.init` is not called immediately. in that case the difference would be higher.
So yes, the static variable is enough

**Review by stefanosiano — APPROVED** [review]
nice!

**Pull request merged**